aquadoggo/src/graphql/replication/single_entry_and_payload.rs

```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use async_graphql::*;
+
+use crate::db::stores::StorageEntry;
+use crate::graphql::Context as GraphQLContext;
+
+use super::payload::Payload;
+use super::Entry;
+
+/// A p2panda entry with optional payload and the collection of skiplinks required to verify it.
+#[derive(Debug)]
+pub struct SingleEntryAndPayload {
+    pub entry: Entry,
+    pub payload: Option<Payload>,
+}
+
+#[Object]
+impl SingleEntryAndPayload {
+    /// The entry
+    async fn entry(&self) -> &Entry {
+        &self.entry
+    }
+
+    /// The payload
+    async fn payload(&self) -> Option<&Payload> {
+        self.payload.as_ref()
+    }
+
+    /// Get all the skiplinks for this entry that are required to verify the entry is valid
+    async fn skiplinks<'a>(&self, ctx: &Context<'a>) -> Result<Vec<Entry>> {
+        let ctx: &GraphQLContext = ctx.data()?;
+
+        let result = ctx
+            .replication_context
+            .lock()
+            .await
+            .get_skiplinks(&self.entry)
+            .await?;
+
+        Ok(result)
+    }
+}
+
+impl From<StorageEntry> for SingleEntryAndPayload {
+    fn from(entry_row: StorageEntry) -> Self {
+        let entry = Entry(entry_row.entry_signed().to_owned());
+        let payload = entry_row
+            .operation_encoded()
+            .map(|encoded| Payload(encoded.to_owned()));
+        Self { entry, payload }
+    }
+}
```

Comment:
Could these two structs `StorageEntry` and `SingleEntryAndPayload` be combined into one? `StorageEntry` is really just a wrapper around an entry and it's encoded operation:

```rust
#[derive(Debug, Clone, PartialEq)]
pub struct StorageEntry {
    entry_signed: EntrySigned,
    operation_encoded: OperationEncoded,
}
```

Reply:
No, I intentionally separate them.

`SingleEntryAndPayload` is the type you get back when you do partial replication. It has a gql method that lets you request the `skiplinks`.

I wanted to keep this apart from an "ordinary" entry that would be part of a collection you'd get back from calling eg `get_entries_newer_than_seq`.

It's some defensive programming. Imagine if you could write a query like this:

```gql
query GetEntriesNewerThanSeq($logId: LogId!, $author: Author!, $sequenceNumber: SequenceNumber!, $first: Int, $after: String) {
    getEntriesNewerThanSeq(logId: $logId, author: $author, sequenceNumber: $sequenceNumber, first: $first, after: $after){
        pageInfo {
            hasNextPage
        }
        edges {
            cursor,
            node {
                entry,
                payload,
                skiplinks
            }
        }
    }
}
```

Asking for the skiplinks here would mean it would have to resolve the skip links for every single entry it returns. It would be a handy way to ddos the server.

Reply:
Ahh! Thanks for explaining :+1: